src/test/java/com/google/firebase/internal/RetryUnsuccessfulResponseHandlerTest.java

```diff
@@ -0,0 +1,281 @@
+
+    try {
+      request.execute();
+      fail("No exception thrown for HTTP error");
+    } catch (HttpResponseException e) {
+      assertEquals(404, e.getStatusCode());
+    }
+
+    assertEquals(0, sleeper.getCount());
+    assertEquals(1, failingRequest.getCount());
+  }
+
+  @Test
+  public void testRetryOnHttpClientErrorWhenSpecified() throws IOException {
+    MultipleCallSleeper sleeper = new MultipleCallSleeper();
+    RetryUnsuccessfulResponseHandler handler = new RetryUnsuccessfulResponseHandler(
+        testRetryConfig(sleeper));
+    CountingLowLevelHttpRequest failingRequest = CountingLowLevelHttpRequest.fromStatus(429);
+    HttpRequest request = TestUtils.createRequest(failingRequest);
+    request.setUnsuccessfulResponseHandler(handler);
+    request.setNumberOfRetries(MAX_RETRIES);
+
+    try {
+      request.execute();
+      fail("No exception thrown for HTTP error");
+    } catch (HttpResponseException e) {
+      assertEquals(429, e.getStatusCode());
+    }
+
+    assertEquals(MAX_RETRIES, sleeper.getCount());
+    assertArrayEquals(new long[]{500, 1000, 2000, 4000}, sleeper.getDelays());
+    assertEquals(MAX_RETRIES + 1, failingRequest.getCount());
+  }
+
+
+  @Test
+  public void testRetryAfterIsAbsent() throws IOException {
+    MultipleCallSleeper sleeper = new MultipleCallSleeper();
+    RetryUnsuccessfulResponseHandler handler = new RetryUnsuccessfulResponseHandler(
+        testRetryConfig(sleeper));
+    CountingLowLevelHttpRequest failingRequest = CountingLowLevelHttpRequest.fromStatus(503);
+    HttpRequest request = TestUtils.createRequest(failingRequest);
+    request.setUnsuccessfulResponseHandler(handler);
+    request.setNumberOfRetries(MAX_RETRIES);
+
+    try {
+      request.execute();
+      fail("No exception thrown for HTTP error");
+    } catch (HttpResponseException e) {
+      assertEquals(503, e.getStatusCode());
+    }
+
+    assertEquals(MAX_RETRIES, sleeper.getCount());
+    assertArrayEquals(new long[]{500, 1000, 2000, 4000}, sleeper.getDelays());
+    assertEquals(MAX_RETRIES + 1, failingRequest.getCount());
+  }
+
+  @Test
+  public void testExponentialBackOffDoesNotExceedMaxInterval() throws IOException {
+    MultipleCallSleeper sleeper = new MultipleCallSleeper();
+    RetryUnsuccessfulResponseHandler handler = new RetryUnsuccessfulResponseHandler(
+        testRetryConfig(sleeper));
+    CountingLowLevelHttpRequest failingRequest = CountingLowLevelHttpRequest.fromStatus(503);
+    HttpRequest request = TestUtils.createRequest(failingRequest);
+    request.setUnsuccessfulResponseHandler(handler);
+    request.setNumberOfRetries(10);
+
+    try {
+      request.execute();
+      fail("No exception thrown for HTTP error");
+    } catch (HttpResponseException e) {
+      assertEquals(503, e.getStatusCode());
+    }
+
+    assertEquals(10, sleeper.getCount());
+    assertArrayEquals(
+        new long[]{500, 1000, 2000, 4000, 8000, 16000, 32000, 64000, 120000, 120000},
```

Comment:
There are quite a few tests in this PR that test the same feature at different levels (see testExponentialBackOff). You can certainly leave this as is, but it might be worth considering to reduce duplication in future PRs.